src/spectral.ts

```diff
@@ -143,74 +142,65 @@ export class Spectral {
   }
 
-  /**
-   * Functions
-   */
-
-  public get functions(): FunctionCollection {
-    return this._functions;
-  }
-
+  @deprecated('loadRuleset', '4.1')
   public addFunctions(functions: FunctionCollection) {
-    Object.assign(this._functions, merge({}, functions));
+    this._addFunctions(functions);
   }
 
-  public async loadRuleset(...uris: string[]) {
-    this._addRules(await readRulesFromRulesets(...uris));
+  public _addFunctions(functions: FunctionCollection) {
+    Object.assign(this.functions, functions);
   }
 
-  /**
-   * Rules
-   */
-
-  public get rules(): RunRuleCollection {
-    const rules: RunRuleCollection = {};
-
-    for (const name in this._rules) {
-      if (!this._rules.hasOwnProperty(name)) continue;
-      const rule = this._rules[name];
-
-      rules[name] = {
-        name,
-        ...rule,
-        severity: rule.severity === undefined ? DEFAULT_SEVERITY_LEVEL : getDiagnosticSeverity(rule.severity),
-      };
+  public setFunctions(functions: FunctionCollection) {
+    for (const key in this.functions) {
+      if (!Object.hasOwnProperty.call(this.functions, key)) continue;
+      delete this.functions[key];
     }
 
-    return rules;
+    this._addFunctions({ ...functions });
   }
 
   @deprecated('loadRuleset', '4.1')
   public addRules(rules: RuleCollection) {
     this._addRules(rules);
   }
 
+  public setRules(rules: RuleCollection) {
+    for (const key in this.rules) {
+      if (!Object.hasOwnProperty.call(this.rules, key)) continue;
+      delete this.rules[key];
+    }
+
+    this._addRules({ ...rules });
+  }
+
   private _addRules(rules: RuleCollection) {
-    Object.assign(this._rules, merge({}, rules));
+    for (const name in rules) {
+      if (!rules.hasOwnProperty(name)) continue;
+      const rule = rules[name];
+
+      this.rules[name] = {
+        name,
+        ...rule,
+        severity: rule.severity === void 0 ? DEFAULT_SEVERITY_LEVEL : getDiagnosticSeverity(rule.severity),
+      };
+    }
   }
 
   public mergeRules(rules: PartialRuleCollection) {
-    for (const ruleName in merge({}, rules)) {
-      if (!rules.hasOwnProperty(ruleName)) continue;
-      const rule = rules[ruleName];
+    for (const name in rules) {
+      if (!rules.hasOwnProperty(name)) continue;
+      const rule = rules[name];
       if (rule) {
-        this._rules[ruleName] = merge(this._rules[ruleName], rule);
+        this.rules[name] = merge(this.rules[name], rule);
       }
     }
   }
 
-  public applyRuleDeclarations(declarations: RuleDeclarationCollection) {
-    for (const ruleName in declarations) {
-      if (!declarations.hasOwnProperty(ruleName)) continue;
-      const declaration = declarations[ruleName];
-
-      const rule = this.rules[ruleName];
-      if (rule) {
-        if (typeof declaration === 'boolean') {
-          this._rules[ruleName].recommended = declaration;
-        }
-      }
-    }
+  public async loadRuleset(...uris: string[]) {
```

Comment:
This is a very constrictive API - we'll never be able to change it or add any options without introducing a breaking change. I suggest something like this instead:

```ts
public async loadRulesets(uris: string[]) {
```

With this, we can easily add an optional `opts` or other argument after uris in the future, as needed. I also pluralized the method, since it supports multiple.

Reply:
Yeah, you are right. I'll pull over the code I wrote herehttps://github.com/stoplightio/spectral/pull/460 it has a very similar APIs. I had to change it because of the reasons you mentioned.